### PR TITLE
dumpling: fix incorrect progress report when dump with --sql 

### PR DESCRIPTION
**Pull request by docsir**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/30532

Problem Summary: dump with --sql has incorrect progress report

### What is changed and how it works?
fix incorrect progress report including `estimate total rows` and `tables`
### Check List

Tests 

- [ ] Unit test(it will be added later)

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dumpling/export/dump.go [hunk 1]
```diff
@@ -1137,10 +1137,15 @@ func dumpTableMeta(conf *Config, conn *sql.Conn, db string, table *TableInfo) (T
 }
 
 func (d *Dumper) dumpSQL(tctx *tcontext.Context, taskChan chan<- Task) {
+	var conn *sql.Conn
 	conf := d.conf
 	meta := &tableMeta{}
 	data := newTableData(conf.SQL, 0, true)
 	task := NewTaskTableData(meta, data, 0, 1)
+	fieldName, _ := pickupPossibleField(meta, conn)
+	c := estimateCount(tctx, meta.DatabaseName(), meta.TableName(), conn, fieldName, conf)
```
Comment (lichunzhu) [inline dumpling/export/dump.go:1146]: `--sql` doesn't have either `meta.DatabaseName()` or `meta.TableName()`

File: dumpling/export/dump.go [hunk 2]
```diff
@@ -1137,10 +1137,14 @@ func dumpTableMeta(conf *Config, conn *sql.Conn, db string, table *TableInfo) (T
 }
 
 func (d *Dumper) dumpSQL(tctx *tcontext.Context, taskChan chan<- Task) {
+	var conn *sql.Conn
```
Comment (lichunzhu) [inline dumpling/export/dump.go:1140]: This conn is a nil pointer. Should use outside `metaConn` for this.

File: dumpling/export/dump.go [hunk 3]
```diff
@@ -1136,11 +1136,14 @@ func dumpTableMeta(conf *Config, conn *sql.Conn, db string, table *TableInfo) (T
 	return meta, nil
 }
 
-func (d *Dumper) dumpSQL(tctx *tcontext.Context, taskChan chan<- Task) {
+func (d *Dumper) dumpSQL(tctx *tcontext.Context, metaConn *sql.Conn, taskChan chan<- Task) {
 	conf := d.conf
 	meta := &tableMeta{}
 	data := newTableData(conf.SQL, 0, true)
 	task := NewTaskTableData(meta, data, 0, 1)
+	c := detectEstimateRows(tctx, metaConn, fmt.Sprintf("EXPLAIN `%s`", conf.SQL), []string{"rows", "estRows", "count"})
```
Comment (lichunzhu) [inline dumpling/export/dump.go:1144]: don't use quotations for sql
```suggestion
	c := detectEstimateRows(tctx, metaConn, fmt.Sprintf("EXPLAIN %s", conf.SQL), []string{"rows", "estRows", "count"})
```

**Review by lichunzhu — APPROVED** [review]

**Review by glorv — APPROVED** [review]

**Review by Ehco1996 — APPROVED** [review]

**Pull request merged**